File: packages/cli/src/ckpt/storage/local.py
"""LocalStore — SQLite CRUD for reasoning and staging records."""

from __future__ import annotations

import json
import sqlite3
import uuid
from datetime import datetime, timezone

from ckpt.core.config import ensure_ckpt_dir, get_db_path
from ckpt.storage.schema import migrate


class LocalStore:
    """SQLite-backed local storage for reasoning data."""

    def __init__(self) -> None:
        ensure_ckpt_dir()
        self._db_path = get_db_path()
        self._conn: sqlite3.Connection | None = None

    @property
    def conn(self) -> sqlite3.Connection:
        if self._conn is None:
            self._conn = sqlite3.connect(str(self._db_path))
            self._conn.execute("PRAGMA journal_mode=WAL")
            self._conn.row_factory = sqlite3.Row
            migrate(self._conn.cursor())
            self._conn.commit()
        return self._conn

    def close(self) -> None:
        if self._conn:
            self._conn.close()
            self._conn = None

    # --- Staging ---

    def add_staging(
        self,
        files: list[str],
        reasoning: str | None = None,
        metadata: dict | None = None,
    ) -> str:
        """Create a staging record. Returns the record ID."""
        record_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc).isoformat()
        self.conn.execute(
            "INSERT INTO staging (id, reasoning, files, timestamp, metadata) VALUES (?, ?, ?, ?, ?)",
            (
                record_id,
                reasoning,
                json.dumps(files),
                now,
                json.dumps(metadata or {}),
            ),
        )
        self.conn.commit()
        return record_id

    def get_staging_records(self) -> list[dict]:
        """Return all staging records."""
        rows = self.conn.execute("SELECT * FROM staging ORDER BY timestamp").fetchall()
        return [dict(row) for row in rows]

    def get_staging_reasoning(self) -> str | None:
        """Aggregate reasoning from all staging records."""
        rows = self.conn.execute(
            "SELECT reasoning FROM staging WHERE reasoning IS NOT NULL ORDER BY timestamp"
        ).fetchall()
        parts = [row["reasoning"] for row in rows if row["reasoning"]]
        return "\n".join(parts) if parts else None

    def clear_staging(self) -> None:
        """Delete all staging records."""
        self.conn.execute("DELETE FROM staging")
        self.conn.commit()

    # --- Reasoning ---

    def add_reasoning(
        self,
        commit_hash: str,
        reasoning: str | None,
        author: str,
        files: list[str],
        parent_hash: str = "",
        metadata: dict | None = None,
    ) -> str:
        """Create a reasoning record. Returns the record ID."""
        record_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc).isoformat()
        self.conn.execute(
            """INSERT INTO reasoning
               (id, commit_hash, reasoning, author, timestamp, files, parent_hash, metadata, synced, created_at, updated_at)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?)""",
            (
                record_id,
                commit_hash,
                reasoning,
                author,
                now,
                json.dumps(files),
                parent_hash,
                json.dumps(metadata or {}),
                now,
                now,
            ),
        )
        self.conn.commit()
        return record_id

    def get_reasoning_by_commit(self, commit_hash: str) -> dict | None:
        """Look up reasoning for a given commit hash."""
        row = self.conn.execute(
            "SELECT * FROM reasoning WHERE commit_hash = ?", (commit_hash,)
        ).fetchone()
        return dict(row) if row else None

    def get_unsynced(self) -> list[dict]:
        """Return all reasoning records not yet synced."""
        rows = self.conn.execute(
            "SELECT * FROM reasoning WHERE synced = 0 ORDER BY timestamp"
        ).fetchall()
        return [dict(row) for row in rows]

    def mark_synced(self, record_ids: list[str]) -> None:
        """Mark records as synced."""
        if not record_ids:
            return
        now = datetime.now(timezone.utc).isoformat()
        placeholders = ",".join("?" for _ in record_ids)
        self.conn.execute(
            f"UPDATE reasoning SET synced = 1, updated_at = ? WHERE id IN ({placeholders})",
            [now, *record_ids],
        )
        self.conn.commit()

    def get_reasoning_records(self, limit: int = 50) -> list[dict]:
        """Return recent reasoning records."""
        rows = self.conn.execute(
            "SELECT * FROM reasoning ORDER BY timestamp DESC LIMIT ?", (limit,)
        ).fetchall()
        return [dict(row) for row in rows]

File: packages/cli/src/storage/local.ts
import Database from 'better-sqlite3';
import { existsSync } from 'fs';
import { randomUUID } from 'crypto';
import { getCkptDir, getDbPath } from '../core/config.js';
import { migrate } from './schema.js';

export class LocalStore {
  private db: Database.Database;

  constructor() {
    const dir = getCkptDir();
    if (!existsSync(dir)) {
      console.error('Not a ckpt directory. Run `ckpt init` first.');
      process.exit(1);
    }
    this.db = new Database(getDbPath());
    this.db.pragma('journal_mode = WAL');
    migrate(this.db);
  }

  close(): void {
    this.db.close();
  }

  // --- Staging ---

  addStaging(files: string[], reasoning?: string, metadata?: Record<string, unknown>): string {
    const id = randomUUID();
    const now = new Date().toISOString();
    this.db
      .prepare('INSERT INTO staging (id, reasoning, files, timestamp, metadata) VALUES (?, ?, ?, ?, ?)')
      .run(id, reasoning ?? null, JSON.stringify(files), now, JSON.stringify(metadata ?? {}));
    return id;
  }

  getStagingRecords(): Record<string, unknown>[] {
    return this.db
      .prepare('SELECT * FROM staging ORDER BY timestamp')
      .all() as Record<string, unknown>[];
  }

  getStagingReasoning(): string | undefined {
    const rows = this.db
      .prepare('SELECT reasoning FROM staging WHERE reasoning IS NOT NULL ORDER BY timestamp')
      .all() as { reasoning: string }[];
    const parts = rows.map((r) => r.reasoning).filter(Boolean);
    return parts.length > 0 ? parts.join('\n') : undefined;
  }

  clearStaging(): void {
    this.db.prepare('DELETE FROM staging').run();
  }

  // --- Reasoning ---

  addReasoning(
    commitHash: string,
    reasoning: string | undefined,
    author: string,
    files: string[],
    parentHash: string = '',
    metadata?: Record<string, unknown>,
    repoUrl?: string
  ): string {
    const id = randomUUID();
    const now = new Date().toISOString();
    this.db
      .prepare(
        `INSERT INTO reasoning
         (id, commit_hash, reasoning, author, timestamp, files, parent_hash, metadata, synced, repo_url, created_at, updated_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?, ?)`
      )
      .run(
        id,
        commitHash,
        reasoning ?? null,
        author,
        now,
        JSON.stringify(files),
        parentHash,
        JSON.stringify(metadata ?? {}),
        repoUrl ?? null,
        now,
        now
      );
    return id;
  }

  getReasoningByCommit(commitHash: string): Record<string, unknown> | undefined {
    return this.db
      .prepare('SELECT * FROM reasoning WHERE commit_hash = ?')
      .get(commitHash) as Record<string, unknown> | undefined;
  }

  getUnsynced(): Record<string, unknown>[] {
    return this.db
      .prepare('SELECT * FROM reasoning WHERE synced = 0 ORDER BY timestamp')
      .all() as Record<string, unknown>[];
  }

  markSynced(recordIds: string[]): void {
    if (recordIds.length === 0) return;
    const now = new Date().toISOString();
    const placeholders = recordIds.map(() => '?').join(',');
    this.db
      .prepare(`UPDATE reasoning SET synced = 1, updated_at = ? WHERE id IN (${placeholders})`)
      .run(now, ...recordIds);
  }

  getReasoningRecords(limit: number = 50): Record<string, unknown>[] {
    return this.db
      .prepare('SELECT * FROM reasoning ORDER BY timestamp DESC LIMIT ?')
      .all(limit) as Record<string, unknown>[];
  }
}
